lib/active_merchant/billing/gateways/global_collect.rb

```diff
@@ -307,11 +307,14 @@ def json_error(raw_response)
       end
 
       def headers(action, post, authorization = nil)
```

Comment:
`options` is not being passed to this method. So `options[:idempotency_key]` is always going to be `nil` here.

Part of what's confusing about this is that you would expect `options[:idempotency_key]` on line 316 to throw an error if `options` doesn't exist. But what's happening is that `options` has _also_ been declared as an instance variable with read-only access in the Gateway class [here](https://github.com/activemerchant/active_merchant/blob/master/lib/active_merchant/billing/gateway.rb#L145), and since every gateway adapter inherits from that class, there is an `options` variable that floats around in our gateway adapters and causes problems in situations like this.

If you want to see this in action, I recommend using a `binding.pry` breakpoint both in this method and in `def authorize`, then running your unit test and seeing what `options` returns in each scope.

This issue with `options` as an invisible instance variable is a perennial source of pain that we are kind of just stuck with, unfortunately.

Anyways, this method needs access to the local `options` variable so that the `idempotency_key` can be added, if present.

Reply:
@dsmcclain Thank you for the thorough review. I've pushed up changes! They ended up being significant due to some 'aha' moments I had when making corrections.